test/distributed/algorithms/ddp_comm_hooks/test_ddp_hooks.py

```diff
@@ -0,0 +1,170 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import os
+
+import numpy as np
+import torch
+import torch.distributed as c10d
+from torch import nn
+from torch.distributed.algorithms.ddp_comm_hooks import register_ddp_comm_hook
+from torch.nn.parallel import DistributedDataParallel
+from torch.testing._internal.common_distributed import (
+    MultiProcessTestCase,
+    requires_nccl,
+    skip_if_lt_x_gpu,
+)
+from torch.testing._internal.common_utils import run_tests
+
+
+def gpus_for_rank(world_size):
+    visible_devices = list(range(torch.cuda.device_count()))
+    gpus_per_process = torch.cuda.device_count() // world_size
+    gpus_for_rank = []
+    for rank in range(world_size):
+        gpus_for_rank.append(
+            visible_devices[rank * gpus_per_process : (rank + 1) * gpus_per_process]
+        )
+    return gpus_for_rank
+
+
+class Task(nn.Module):
+    def __init__(self):
+        super(Task, self).__init__()
+        torch.manual_seed(0)
+        self.p = nn.Parameter(torch.randn(40, 20))
+
+    def forward(self, x):
+        return self.p * x
+
+
+class TestDdpCommHook(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.t0 = Task()
+
+    def forward(self, x, rank):
+        return self.t0(x ** (1 + rank))
+
+
+class DistributedDataParallelCommHookTest(MultiProcessTestCase):
+    def setUp(self):
+        super(DistributedDataParallelCommHookTest, self).setUp()
+        self._fork_processes()
+
+    def tearDown(self):
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    @property
+    def world_size(self):
+        return 2
+
+    def _local_model(self):
+        local_model = TestDdpCommHook().cpu()
+
+        return local_model
+
+    def _get_grads(self, process_group, hook=None):
+        device_id = gpus_for_rank(self.world_size)[self.rank][0]
+        gpu_model = DistributedDataParallel(
+            TestDdpCommHook().to(device_id),
+            device_ids=[device_id],
+            process_group=process_group,
+        )
+
+        # Register DDP Communication Hook if defined
+        if hook is not None:
+            register_ddp_comm_hook(
+                comm_hook_name=hook, model=gpu_model, state=process_group
```

Comment:
Does this work correctly even though `comm_hook_name` is a string and `register_ddp_comm_hook` needs an Enum as input? It might be nicer to pass in the Enum here itself instead of relying on some string to Enum implicit conversion.

Reply:
That's correct, thanks for reminding this.